src/main/java/com/mongodb/hibernate/jdbc/MongoPreparedStatement.java

```diff
@@ -0,0 +1,360 @@
+        setBsonDateTimeParameter(parameterIndex, x, Types.TIME);
+    }
+
+    @Override
+    public void setTimestamp(int parameterIndex, @Nullable Timestamp x, @Nullable Calendar cal) throws SQLException {
+        checkClosed();
+        checkParameterIndex(parameterIndex);
+        setBsonDateTimeParameter(parameterIndex, x, Types.TIMESTAMP);
+    }
+
+    @Override
+    public void setNull(int parameterIndex, int sqlType, @Nullable String typeName) throws SQLException {
+        checkClosed();
+        checkParameterIndex(parameterIndex);
+        throw new NotYetImplementedException("To be implemented during Array / Struct tickets");
+    }
+
+    @SuppressWarnings("JavaUtilDate")
+    private void setBsonDateTimeParameter(int parameterIndex, java.util.@Nullable Date date, int sqlType)
+            throws SQLException {
+        if (date == null) {
+            setNull(parameterIndex, sqlType);
+        } else {
+            setParameter(parameterIndex, new BsonDateTime(date.getTime()));
+        }
+    }
+
+    private void setParameter(int parameterIndex, BsonValue parameterValue) {
+        var parameterValueSetter = parameterValueSetters.get(parameterIndex - 1);
+        parameterValueSetter.accept(parameterValue);
+    }
+
+    private static void parseParameters(BsonDocument command, List<Consumer<BsonValue>> parameterValueSetters) {
+        for (var entry : command.entrySet()) {
+            if (isParameterMarker(entry.getValue())) {
+                parameterValueSetters.add(entry::setValue);
+            } else if (entry.getValue().getBsonType().isContainer()) {
+                parseParameters(entry.getValue(), parameterValueSetters);
+            }
+        }
+    }
+
+    private static void parseParameters(BsonArray array, List<Consumer<BsonValue>> parameterValueSetters) {
+        for (var i = 0; i < array.size(); i++) {
+            var value = array.get(i);
+            if (isParameterMarker(value)) {
+                var idx = i;
+                parameterValueSetters.add(v -> array.set(idx, v));
+            } else if (value.getBsonType().isContainer()) {
+                parseParameters(value, parameterValueSetters);
+            }
+        }
+    }
+
+    private static void parseParameters(BsonValue value, List<Consumer<BsonValue>> parameterValueSetters) {
+        if (value.isDocument()) {
+            parseParameters(value.asDocument(), parameterValueSetters);
+        } else if (value.isArray()) {
+            parseParameters(value.asArray(), parameterValueSetters);
+        } else {
+            fail("Only BSON container type (BsonDocument or BsonArray) is accepted");
+        }
+    }
+
+    private static boolean isParameterMarker(BsonValue value) {
+        return value.getBsonType() == BsonType.UNDEFINED;
+    }
+
+    private void checkParameterIndex(int parameterIndex) throws SQLException {
+        if (parameterValueSetters.isEmpty()) {
+            throw new SQLException("No parameter exists");
```

Comment:
I found an edge. case above. If the parameter list is empty, the below exception message would be invalid.